### ISSUE TITLE
Citation Label

### ISSUE THREAD

**Issue by rmzelle** [citation-style-language/zotero-bits#24]
By adding a "Citation Label"/**citationLabel** field (mapped  to **citation-label** in CSL) to all item types, we should be able to support most label styles. Users will have to add the labels themselves, although here Zotero's forthcoming read/write API might help out: it would be relatively easy to write a label generator.

**Comment by AllThatIsTheCase** [comment]
Is any news available on a possible ETA for that revamp?

**Comment by adam3smith** [comment]
"later this year" has been what Dan is saying and recent activity on github suggests things are pretty far advanced.

**Comment by bwiernik** [comment]
Hmm, note that there are two things being described as a "citation label". 

The original post referred to `citation-label`--a variable that should actually appear in text for label/trigraph styles (e.g., "SmJo1965"). There is a discussion for providing a syntax for automatically generating labels [here](https://github.com/citation-style-language/csl-evolution/issues/30). 

In all of the replies, that variable is usually described as a "citation key" or "citekey". That is analogous to the `id` variable in CSL-JSON and indicates a placeholder used while writing in plain text systems like BibTeX or Markdown, which is eventually replaced with formatted citations.

**Comment by dstillman** [comment]
So do we want `citationKey` (for BibTeX/etc.) and `citationLabel` (for label styles) for all types, or can we just have a `citationKey` field that also gets used for label styles? (People are going to be confused by the difference.)

**Comment by retorquere** [comment]
I don't know what citation-label does in CSL, but if there's a strong overlap, I could switch to citationLabel with citationKey as an alias. Bibtex users will be more used to citation key though, and semantically, in bibtex, it is a key (it uniquely identifies an item) rather than a label (which describes one of potentially multiple characteristics of an item, and can be shared between items).

**Comment by adam3smith** [comment]
My view is the field should just be key. Label formats actually vary by citation style, so ideally would be configurable algorithmically, but they are also fairly rare.

**Comment by retorquere** [comment]
WRT algorithmic generation, would appreciate being part of the dialogue, I'd prefer to align BBTs pattern format as close as possible (even if only by adding a formatter function that will yield a style-bound key generated by the csl processor - people are going to want them)

**Comment by bwiernik** [comment]
CSL 1.1 will include style syntax for specifying a label. citation-label will be generated by the processor and should not have an item field in Zotero. 

citation-key ala BibTeX should be specified as part of the item. This is mostly to facilitate use with BibTeX/pandoc systems to be a human-readable variable that can be called in lieu of the machine-readable id

**Comment by retorquere** [comment]
@dstillman anything I should be taking note of for this change?

**Comment by bwiernik** [comment]
@retorquere with this change, every item type now will have a native `citationKey` field that BBT can interface with (other than notes and attachments, which don't have most fields generally shouldn't be cited)

**Comment by retorquere** [comment]
Is this projected to be released in the 8.0 line, or already in 7.0?

I'd like to have a conversation on how BBT should interact with this field. Writing to this field would be equivalent to what BBT pinning does now, but pinning is not the default BBT behavior, and if I make pinning in the citekey field (when empty of course) the default I might step on toes. Looking to do the right thing here.

**Comment by dstillman** [comment]
This will be in Z8, but Z7 will likely either get an update with these new fields at the same time or be auto-pushed to Z8 soon after Z8's release, since it introduces sync incompatibility.

For now, we won't be doing anything special with this field other than adding it to all item types. It's been on the two most recently added types (Dataset and Standard) for a while, so it's no different from what the situation has been for those fields.

We can talk about any special behavior we might want to do separately.

**Comment by bwiernik** [comment]
Showing the autogenerated cite key in grey text and then having a pin button and copy button next to it would be a really pleasant experience

**Comment by retorquere** [comment]
Is that something I can do with the item field API as it is now?

**Comment by dstillman** [comment]
No, sorry. For now, you'll just need to keep using your custom field at the top and sync that to the real Citation Key field. But since Citation Key will now appear in the list of fields for all items, you might consider hiding that with [some custom CSS](https://github.com/zotero/make-it-red/blob/70f709d1a330e9c572b2ebf81ba31d2eab8dd390/src-2.0/make-it-red.js#L23-L30).

How much functionality is there in BBT regarding citation key generation/maintenance? Is this something that realistically makes sense as part of Zotero, if people should be using BBT for BibTeX functionality anyway?

**Comment by retorquere** [comment]
It's a little lopsided - the base is very compact but then there is also a fair bit of stuff that wouldn't make sense to have in Zotero. The dictionaries for the CJK stuff for example is currently the bulk of the shipped size of BBT - that could easily be split off to a separate plugin, but I just didn't feel like dealing with cross - plugin consistency problems. I load it dynamically to save memory but it's shipped all the same.

The formula compiler is pretty compact and has very little coupling - functions and filters and their verifiers can in principle be fed to it dynamically, and it returns a javascript function that takes a zotero item and returns a citation key. A base set of functions/filters could be set by zotero, BBT could augment those; I could make a sensible subset TBD and split that off. Disambiguation is separate from the compiler.

**Comment by retorquere** [comment]
WIth the native citation key field, I don't think non-pinned keys fit in the UX. I don't think it is a good fit for for the general behavior of native zotero fields, and I don't think BBT should ghost-fill the field with a key that doesn't sync. So for Zotero 8, I propose BBT lets go of the concept in toto. This would leave two options (either fixed or configurable):

* When BBT is active it behaves as auto-pin does now; after a configurable amount of time, BBT fills out the field iff it is empty
* BBT does nothing automatically but offers a "fill now" option in the context menu

In either case there would be the problem that it is possible to start an export while the field is still empty. For the stock BibTeX export I would imagine that the existing key generator kicks in at time of export as it does now. For BBT I haven't yet decided what the behavior will be.

I am going to keep the existing behavior (pinned/non-pinned, BBT-created citationkey field in the item pane) for as long as Zotero 7 is the release version, so that will also show up in Z8. Is there an ETA for Z8 release?

**Comment by bwiernik** [comment]
> When BBT is active it behaves as auto-pin does now; after a configurable amount of time, BBT fills out the field iff it is empty

I like this behavior better. 

Two questions:

1. Would it be possible to implement auto-updating if the pattern is changed, such as changing keys that match the previous pattern? 
2. Is it possible to implement auto-updating if metadata is updated?

**Comment by retorquere** [comment]
> 1. Would it be possible to implement auto-updating if the pattern is changed, such as changing keys that match the previous pattern?

Possible, yes, but I removed that behavior long ago (to the consternation of some TBF), with an eye on key stability, and when it did auto-update, it did not affect pinned keys. Right now it requires a refresh by the user to have a new pattern go into effect for unpinned keys. With the loss of the pinned/unpinned keys distinction, I don't think BBT ought to apply a formula update in bulk.

> 2. Is it possible to implement auto-updating if metadata is updated?

It would be possible to make configurable whether native citation keys should all be considered pinned, or unpinned (default: pinned). That does sound like a decent compromise.
